observer.go

```diff
@@ -0,0 +1,69 @@
+package jaeger
+
+import opentracing "github.com/opentracing/opentracing-go"
+
+// Observer can be registered with the Tracer to receive notifications about new Spans.
+type Observer interface {
+	OnStartSpan(operationName string, options opentracing.StartSpanOptions) SpanObserver
+}
+
+// SpanObserver is created by the Observer and receives notifications about other Span events.
+type SpanObserver interface {
+	OnSetOperationName(operationName string)
+	OnSetTag(key string, value interface{})
+	OnFinish(options opentracing.FinishOptions)
+}
+
+// observer is a disatcher to other observers
+type observer struct {
+	observers []Observer
+}
+
+// spanObserver is a dispatcher to other span observers
+type spanObserver struct {
+	observers []SpanObserver
+}
+
+// noopSpanObserver is used when there are no observers registered on the Tracer
+// or none of them returns span observers from OnStartSpan.
+var noopSpanObserver = spanObserver{}
+
+func (o *observer) append(observer Observer) {
+	o.observers = append(o.observers, observer)
+}
+
+func (o observer) OnStartSpan(operationName string, options opentracing.StartSpanOptions) SpanObserver {
+	var spanObservers []SpanObserver
+	for _, obs := range o.observers {
+		spanObs := obs.OnStartSpan(operationName, options)
+		if spanObs != nil {
+			if spanObservers == nil {
+				spanObservers = make([]SpanObserver, 0, len(o.observers))
```

Comment:
When you use `append()` the first capacity of the slice is 2, then it's doubled when it needs to grow. So it's quite possible that in your approach (a) even more memory is allocated than needed, and what's worse (b) potentially more than one memory allocation happens should the slice needs to be extended.

In my approach you get no more than one allocation, and the memory (a pointer) is only wasted if some observers are not interested in the span. I don't think this memory waste is that critical, the service would need to keep a lot of spans in-flight for this to matter